admin/class-acc_user_importer-admin.php

```diff
@@ -358,148 +374,192 @@ private function proccess_user_data ( $users ) {
+				} else {
+					//email is unique, proceed to create a new record
+					$update_this_user = true;
+					$create_this_user = true;
+					$this->log_dual(" > email not found on any other users");
+					$this->log_dual(" > will create new user account");
 					$new_users[] = $user_contact_id;
+					$new_users_email[] = $user_email ;
 					$user_data["role"] = $default_role;
 				}
 			}
 			
 			//didn't find a user by contact ID, and don't have an email to search for either
 			else {
-				$api_response['log'] .= "&nbsp;&gt; <b>error</b>: no email given, cannot create new user account.<br/>";
-				$this->log_local_output(" > error: no email given, cannot create new user account.");
-				$update_this_user = false;
+				$this->log_dual(" > error: no email given, cannot create new user account.");
 				$update_errors[] = $user['FirstName'] . " " . $user['LastName'] . " (" . $user_contact_id . ")";
 			}
 			
+			//If user already exists, update only if something changed.
+			//Do a bunch of compare to decide if an update is needed.
+			//This logic needs to be undated each time we add a new user field.
+			if ($update_this_user && $user_already_exists) {
 
-			//only run if indicated
-			if ($update_this_user) {
+				$update_this_user = false;
 
-				$wp_user = wp_insert_user( $user_data ) ;
-				
-				//attempt to add user
-				if ( is_wp_error($wp_user) ) {
-					
-					$api_response['log'] .= "&nbsp;&gt; failed to update user<br/>";
-					$api_response['log'] .= "&nbsp;&gt; WP:" . $wp_user->get_error_message() . "<br/>";
-					$this->log_local_output(" > failed to update user");
-					$this->log_local_output(" > WP:" . $wp_user->get_error_message());
+				// Get the user object
+				$user_meta = get_userdata($user_id);
+
+				// Get all the user roles as an array.
+				$user_roles = $user_meta->roles;
+				// Check if user has expired or ex-member role. If so, update role to default
+				if ( in_array( $role_expiry1, $user_roles, true ) ) {
+					$this->log_dual(" > User is " . $role_expiry1 . ", refreshing role to " . $default_role);
+					$update_this_user = true;
+					$user_data["role"] = $default_role;
+					$role_refreshed[] = $user_contact_id;
+					$role_refreshed_email[] = $user_email;
+				} elseif (in_array( $role_expiry2, $user_roles, true ) ) {
+					$this->log_dual(" > User is " . $role_expiry2 . ", refreshing role to " . $default_role);
+					$update_this_user = true;
+					$user_data["role"] = $default_role;
+					$role_refreshed[] = $user_contact_id;
+					$role_refreshed_email[] = $user_email;
+				}
+
+				//Check if email changed
+				if (!$update_this_user) {
+					if ($user_email != $user_meta->user_email) {
+						$this->log_dual(" > email changed from " . $user_meta->user_email . " to " . $user_email);
+						$update_this_user = true;
+					}
+				}
+
+				//Check if HomePhone changed
+				if (!$update_this_user) {
+					if ($user["HomePhone"] != $user_meta->home_phone) {
```

Comment:
Without a check, if the "HomePhone" key exists, this will throw PHP notices whenever it is not existing. Not the end of the world, but it will flood logs and error tracking tools. We also need to consider the case if the metadata does not exist yet, as far as I can tell, then `$user_meta->home_phone` returns an empty string.

To avoid accessing an empty index, you could leverage the [Null coalescing operator](https://www.php.net/manual/en/migration70.new-features.php#migration70.new-features.null-coalesce-op) and write something like this:
```php
if (($homePhone = $user["HomePhone"] ?? '') != $user_meta->home_phone) {
  $this->log_dual(" > home phone changed from " . $user_meta->home_phone . " to " . $homePhone);
  $update_this_user = true;
}
```
That said, ideally, we'd re-use the `$homePhone` variable for updating the value as well, but that would be a bit convoluted as it stands now. I like the logic generally to only update if required, but I think the implementation can be simplified quite a bit. I am fine merging this first, then I give it a shot on top of your implementation, or directly committing to your branch. Your choice :slightly_smiling_face:

Reply:
OK, I will let you do the changes.

Reply:
I did the changes